rllib/core/rl_module/marl_module.py

```diff
@@ -0,0 +1,295 @@
+from typing import Iterator, Mapping, Any, Union, Dict, Set
+import pprint
+
+
+from ray.rllib.utils.annotations import (
+    OverrideToImplementCustomLogic,
+    override,
+)
+
+from ray.rllib.models.specs.specs_dict import ModelSpec
+from ray.rllib.policy.sample_batch import MultiAgentBatch
+from ray.rllib.core.rl_module import RLModule
+
+ModuleID = str
+
+
+class MultiAgentRLModule(RLModule):
+    """Base class for multi-agent RLModules.
+
+    This class holds a mapping from module_ids to RLModules. It provides a convenient
+    way of accessing each individual module as well as all of with only one call.
+
+    The default implementation assumes the data communicated as input and output of
+    RLModule APIs are multi-agent batches. Each RLModule simply takes its own input and
+    produces its own output. It also assumes that the RLModules do not share any
+    parameters or communication with one another. The behavior of modules with such
+    advanced communication would be undefined. To share parameters or communication
+    between RLModules, you should implement your own MultiAgentRLModule.
+
+    Input config keys:
+        `modules`: Mapping from module_id to RLModule config.
+        `trainable_modules`: Set of module_ids that are trainable. If not specified,
+            all modules are trainable.
+
+    RLModule config values can be of the following forms:
+        1. The RLModule instance itself.
+        2. A dict with the following keys:
+            `module_class`: The RLModule class Type or full path separate by `.`.
+                (e.g. `ray.rllib.algoirhms.dqn.torch.DQNTorchRLModule`).
+            `module_config`: The config object for the RLModule.
+        3. A tuple of (RLModule class Type, config object). The RLModule class Type can
+            be the type itself or the full path separate by `.`.
+
+
+    Args:
+        config: The config dict for the multi-agent RLModule (see above).
+    """
+
+    def __init__(self, config: Mapping[str, Any], **kwargs) -> None:
+        super().__init__(config, **kwargs)
+
+        self._rl_modules: Mapping[ModuleID, RLModule] = self._make_modules()
+        self._trainable_rl_modules = self.config.get(
+            "trainable_modules", set(self._rl_modules.keys())
+        )
+
+    def keys(self) -> Iterator[ModuleID]:
+        """Returns an iteratable of module ids."""
+        return self._rl_modules.keys()
+
+    def get_trainable_module_ids(self) -> Set[ModuleID]:
+        """Returns the set of ids of the trainable modules."""
+        return self._trainable_rl_modules
+
+    @override(RLModule)
+    def make_distributed(self, dist_config: Mapping[str, Any] = None) -> None:
+        """Makes the module distributed.
```

Comment:
Can we add more specifics here? What does "distributed" mean exactly?

Reply:
This is for Avnish to implement. For torch it will be wrapping the module in DDP, putting copies on separate devices, etc. while syncing their gradients, these kind of stuff.